### PR TITLE
SWC-2328: introducing the SynapseAlert widget

### PR DESCRIPTION
**Pull request by jay-hodgson**

### CODE REVIEW

File: src/main/java/org/sagebionetworks/web/client/widget/entity/controller/SynapseAlert.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+	 * Can be called when you know that the user must be logged in in order to use the widget. 
+	 */
+	void showMustLogin();
+	
+	/**
+	 * Convenience method to answer if the user is currently logged in.
+	 * @return
+	 */
+	boolean isUserLoggedIn();
+
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/widget/entity/controller/SynapseAlert.java:32]: add resetState()

File: src/main/java/org/sagebionetworks/web/client/widget/entity/controller/SynapseAlertImpl.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+import org.sagebionetworks.web.shared.exceptions.UnknownErrorException;
+
+import com.google.gwt.user.client.rpc.AsyncCallback;
+import com.google.gwt.user.client.ui.Widget;
+import com.google.inject.Inject;
+public class SynapseAlertImpl implements SynapseAlert, SynapseAlertView.Presenter {
+	GlobalApplicationState globalApplicationState;
+	AuthenticationController authController;
+	SynapseAlertView view;
+	SynapseJSNIUtils synapseJSNIUtils;
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/widget/entity/controller/SynapseAlertImpl.java]: remove synapseJSNIUtils

File: src/main/java/org/sagebionetworks/web/client/widget/entity/controller/SynapseAlertImpl.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+			globalApplicationState.getPlaceChanger().goTo(new LoginPlace(LoginPlace.LOGOUT_TOKEN));
+		} else if(ex instanceof ForbiddenException) {			
+			if(!isLoggedIn) {
+				view.showLoginAlert();
+			} else {
+				view.showError(DisplayConstants.ERROR_FAILURE_PRIVLEDGES + " " + ex.getMessage());
+			}
+		} else if(ex instanceof BadRequestException) {
+			//show error (not to file a jira though)
+			view.showError(ex.getMessage());
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/widget/entity/controller/SynapseAlertImpl.java]: No special handling here, skip BadRequestException

File: src/main/java/org/sagebionetworks/web/client/widget/entity/controller/SynapseAlertViewImpl.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+		this.presenter = presenter;
+	}
+	
+	@Override
+	public void showErrorMessage(String message) {
+		DisplayUtils.showErrorMessage(message);
+	}
+
+	@Override
+	public void showConfirmDialog(String title, String string, Callback callback) {
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/widget/entity/controller/SynapseAlertViewImpl.java]: remove

File: src/main/java/org/sagebionetworks/web/client/widget/entity/controller/SynapseAlertViewImpl.java [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+			}
+		});
+	}
+	
+	public void setPresenter(Presenter presenter) {
+		this.presenter = presenter;
+	}
+	
+	@Override
+	public void showErrorMessage(String message) {
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/widget/entity/controller/SynapseAlertViewImpl.java]: rename showErrorDialog()

File: src/main/java/org/sagebionetworks/web/client/widget/entity/controller/SynapseAlertImpl.java [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+				@Override
+				public void onSuccess(Void result) {
+					view.hideJiraDialog();
+					view.showInfo("Report sent", "Thank you!");
+				}
+
+				@Override
+				public void onFailure(Throwable caught) {
+					// failure to create issue!
+					view.showErrorMessage(DisplayConstants.ERROR_GENERIC_NOTIFY+"\n"
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/widget/entity/controller/SynapseAlertImpl.java]: just use showError.

File: src/main/java/org/sagebionetworks/web/client/widget/entity/controller/SynapseAlertImpl.java [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+		} else if (ex instanceof UnknownErrorException) {
+			//An unknown error occurred. 
+			//Exception handling on the backend now throws the reason into the exception message.  Easy!
+			view.showError(ex.getMessage());
+			if (isLoggedIn) {
+				view.showJiraDialog(ex.getMessage());
+			}
+		} else {
+			//not recognized
+			view.showError(ex.getMessage());
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/widget/entity/controller/SynapseAlertImpl.java]: if (ex.getMessage()) is null, empty, or "0" (socket timeout?), show that we did not get a response from the server, could be related to your network connection.

File: src/main/java/org/sagebionetworks/web/client/widget/table/v2/results/TableQueryResultWidget.java [hunk 8]
```diff
@@ -64,7 +70,12 @@ public void configure(Query query, boolean isEditable, QueryResultsListener list
 		this.isEditable = isEditable;
 		this.startingQuery = query;
 		this.queryListener = listener;
-		runQuery();
+		if (!synapseAlert.isUserLoggedIn()) {
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/widget/table/v2/results/TableQueryResultWidget.java:73]: should be removed if backend logic changes.

File: src/test/java/org/sagebionetworks/web/unitclient/widget/entity/controller/SynapseAlertImplTest.java [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+		
+		AsyncMockStubber.callFailureWith(new Exception("ex")).when(mockJiraClient).createIssueOnBackend(anyString(),  any(Throwable.class),  anyString(), any(AsyncCallback.class));
+		String userReport = "clicked a button";
+		widget.onCreateJiraIssue(userReport);
+		verify(mockView, never()).hideJiraDialog();
+		verify(mockView).showErrorMessage(anyString());
+	}
+	
+	@Test
+	public void testHandleServiceUnrecognizedException() {
```
Comment (jay-hodgson) [inline src/test/java/org/sagebionetworks/web/unitclient/widget/entity/controller/SynapseAlertImplTest.java:140]: add tests for empty unrecognized exceptions (empty, ...)

**Comment by john-hill** [comment]
Looks good!
